File: sdks/go/client/protocol/client_message.go
package protocol

import "github.com/clockworklabs/SpacetimeDB/sdks/go/bsatn"

// ClientMessage represents a message sent from client to server.
// It is a BSATN sum type with variants for each message kind.
type ClientMessage interface {
	bsatn.Serializable
	clientMessageTag() uint8
}

// Subscribe requests a new subscription to a set of queries.
// Tag 0 in the ClientMessage sum type.
type Subscribe struct {
	RequestID    uint32
	QuerySetID   uint32
	QueryStrings []string
}

func (*Subscribe) clientMessageTag() uint8 { return 0 }

func (s *Subscribe) WriteBsatn(w bsatn.Writer) {
	w.PutSumTag(0)
	w.PutU32(s.RequestID)
	// QuerySetId is a product with a single u32 field
	w.PutU32(s.QuerySetID)
	// query_strings: Box<[Box<str>]> serialized as array of strings
	w.PutArrayLen(uint32(len(s.QueryStrings)))
	for _, q := range s.QueryStrings {
		w.PutString(q)
	}
}

// UnsubscribeFlags controls the behavior of an Unsubscribe request.
type UnsubscribeFlags uint8

const (
	// UnsubscribeFlagsDefault is the default unsubscribe behavior.
	UnsubscribeFlagsDefault UnsubscribeFlags = 0
	// UnsubscribeFlagsSendDroppedRows requests the server send dropped rows.
	UnsubscribeFlagsSendDroppedRows UnsubscribeFlags = 1
)

// Unsubscribe removes a previously-registered subscription.
// Tag 1 in the ClientMessage sum type.
type Unsubscribe struct {
	RequestID  uint32
	QuerySetID uint32
	Flags      UnsubscribeFlags
}

func (*Unsubscribe) clientMessageTag() uint8 { return 1 }

func (u *Unsubscribe) WriteBsatn(w bsatn.Writer) {
	w.PutSumTag(1)
	w.PutU32(u.RequestID)
	w.PutU32(u.QuerySetID)
	// UnsubscribeFlags is a sum type enum: tag byte only, empty product payload
	w.PutSumTag(uint8(u.Flags))
}

// OneOffQuery runs a query once without subscribing to updates.
// Tag 2 in the ClientMessage sum type.
type OneOffQuery struct {
	RequestID   uint32
	QueryString string
}

func (*OneOffQuery) clientMessageTag() uint8 { return 2 }

func (o *OneOffQuery) WriteBsatn(w bsatn.Writer) {
	w.PutSumTag(2)
	w.PutU32(o.RequestID)
	w.PutString(o.QueryString)
}

// CallReducer invokes a reducer (transactional database function).
// Tag 3 in the ClientMessage sum type.
type CallReducer struct {
	RequestID uint32
	Flags     uint8
	Reducer   string
	Args      []byte
}

func (*CallReducer) clientMessageTag() uint8 { return 3 }

func (c *CallReducer) WriteBsatn(w bsatn.Writer) {
	w.PutSumTag(3)
	w.PutU32(c.RequestID)
	// CallReducerFlags serializes as a plain u8
	w.PutU8(c.Flags)
	w.PutString(c.Reducer)
	// args: Bytes serialized as byte array (u32 len + raw bytes)
	bsatn.WriteByteArray(w, c.Args)
}

// CallProcedure invokes a procedure (non-transactional database function).
// Tag 4 in the ClientMessage sum type.
type CallProcedure struct {
	RequestID uint32
	Flags     uint8
	Procedure string
	Args      []byte
}

func (*CallProcedure) clientMessageTag() uint8 { return 4 }

func (c *CallProcedure) WriteBsatn(w bsatn.Writer) {
	w.PutSumTag(4)
	w.PutU32(c.RequestID)
	// CallProcedureFlags serializes as a plain u8
	w.PutU8(c.Flags)
	w.PutString(c.Procedure)
	// args: Bytes serialized as byte array (u32 len + raw bytes)
	bsatn.WriteByteArray(w, c.Args)
}

File: sdks/swift/Sources/SpacetimeDB/Network/ClientMessage.swift
import Foundation

// MARK: - Client Messages (Encoding)

/// Tag order for Rust `websocket::v2::ClientMessage`:
/// 0 = Subscribe
/// 1 = Unsubscribe
/// 2 = OneOffQuery
/// 3 = CallReducer
/// 4 = CallProcedure
public enum ClientMessage: Encodable {
    case subscribe(Subscribe)
    case unsubscribe(Unsubscribe)
    case oneOffQuery(OneOffQuery)
    case callReducer(CallReducer)
    case callProcedure(CallProcedure)

    public func encode(to encoder: Encoder) throws {}
}

extension ClientMessage: BSATNSpecialEncodable {
    func encodeBSATN(to encoder: _BSATNEncoder) throws {
        switch self {
        case .subscribe(let msg):
            encoder.storage.append(0 as UInt8)
            try msg.encodeBSATN(to: encoder)
        case .unsubscribe(let msg):
            encoder.storage.append(1 as UInt8)
            try msg.encodeBSATN(to: encoder)
        case .oneOffQuery(let msg):
            encoder.storage.append(2 as UInt8)
            try msg.encodeBSATN(to: encoder)
        case .callReducer(let msg):
            encoder.storage.append(3 as UInt8)
            try msg.encodeBSATN(to: encoder)
        case .callProcedure(let msg):
            encoder.storage.append(4 as UInt8)
            try msg.encodeBSATN(to: encoder)
        }
    }
}

/// Rust: `Subscribe { request_id: u32, query_set_id: QuerySetId, query_strings: Box<[Box<str>]> }`
public struct Subscribe: Encodable, BSATNSpecialEncodable {
    public var requestId: RequestId
    public var querySetId: QuerySetId
    public var queryStrings: [String]

    public init(queryStrings: [String], requestId: RequestId, querySetId: QuerySetId = QuerySetId(rawValue: 1)) {
        self.requestId = requestId
        self.querySetId = querySetId
        self.queryStrings = queryStrings
    }

    public func encode(to encoder: Encoder) throws {}

    func encodeBSATN(to encoder: _BSATNEncoder) throws {
        try requestId.encodeBSATN(to: encoder)
        try querySetId.encodeBSATN(to: encoder)
        encoder.storage.append(UInt32(queryStrings.count))
        var singleVal = encoder.singleValueContainer()
        for query in queryStrings {
            try singleVal.encode(query)
        }
    }
}

/// Rust: `Unsubscribe { request_id: u32, query_set_id: QuerySetId, flags: u8 }`
public struct Unsubscribe: Encodable, BSATNSpecialEncodable {
    public var requestId: RequestId
    public var querySetId: QuerySetId
    public var flags: UInt8

    public init(requestId: RequestId, querySetId: QuerySetId, flags: UInt8 = 0) {
        self.requestId = requestId
        self.querySetId = querySetId
        self.flags = flags
    }

    public func encode(to encoder: Encoder) throws {}

    func encodeBSATN(to encoder: _BSATNEncoder) throws {
        try requestId.encodeBSATN(to: encoder)
        try querySetId.encodeBSATN(to: encoder)
        encoder.storage.append(flags)
    }
}

/// Rust: `OneOffQuery { request_id: u32, query_string: Box<str> }`
public struct OneOffQuery: Encodable, BSATNSpecialEncodable {
    public var requestId: RequestId
    public var queryString: String

    public init(requestId: RequestId, queryString: String) {
        self.requestId = requestId
        self.queryString = queryString
    }

    public func encode(to encoder: Encoder) throws {}

    func encodeBSATN(to encoder: _BSATNEncoder) throws {
        try requestId.encodeBSATN(to: encoder)
        var singleVal = encoder.singleValueContainer()
        try singleVal.encode(queryString)
    }
}

/// Rust: `CallReducer { request_id: u32, flags: u8, reducer: Box<str>, args: Bytes }`
public struct CallReducer: Encodable, BSATNSpecialEncodable {
    public var requestId: RequestId
    public var flags: UInt8
    public var reducer: String
    public var args: Data

    public init(requestId: RequestId, flags: UInt8, reducer: String, args: Data) {
        self.requestId = requestId
        self.flags = flags
        self.reducer = reducer
        self.args = args
    }

    public func encode(to encoder: Encoder) throws {}

    func encodeBSATN(to encoder: _BSATNEncoder) throws {
        try requestId.encodeBSATN(to: encoder)
        encoder.storage.append(flags)

        var singleVal = encoder.singleValueContainer()
        try singleVal.encode(reducer)

        encoder.storage.append(UInt32(args.count))
        encoder.storage.append(args)
    }
}

/// Rust: `CallProcedure { request_id: u32, flags: u8, procedure: Box<str>, args: Bytes }`
public struct CallProcedure: Encodable, BSATNSpecialEncodable {
    public var requestId: RequestId
    public var flags: UInt8
    public var procedure: String
    public var args: Data

    public init(requestId: RequestId, flags: UInt8, procedure: String, args: Data) {
        self.requestId = requestId
        self.flags = flags
        self.procedure = procedure
        self.args = args
    }

    public func encode(to encoder: Encoder) throws {}

    func encodeBSATN(to encoder: _BSATNEncoder) throws {
        try requestId.encodeBSATN(to: encoder)
        encoder.storage.append(flags)

        var singleVal = encoder.singleValueContainer()
        try singleVal.encode(procedure)

        encoder.storage.append(UInt32(args.count))
        encoder.storage.append(args)
    }
}
